### PR TITLE
Upgrade ElasticSearch to 7.x

### PR DESCRIPTION
**Pull request by Dillon-Brown**
## Description

- Follow on from https://github.com/salesagility/SuiteCRM/pull/9159 -> https://github.com/salesagility/SuiteCRM/pull/9095 ->https://github.com/salesagility/SuiteCRM/pull/9171.
- Upgrades ElasticSearch to 7.x.

## Mapping types
As mapping types have now been removed from ElasticSearch, I have instead implemented an Index per document type design. Previously we were using a single index ```private $index = 'main';``` along with an accompanying ```defaultParams.yml``` file which specified the mapping.

E.g. Instead of storing accounts and contacts in a single ```main``` index, we now store accounts in the accounts index and contacts in the contacts index. Indices are completely independent of each other and so there will be no conflict of field types between indices.

A few benefits of this design:

- Data is more likely to be dense and so benefit from compression techniques used in Lucene.
- The term statistics used for scoring in full text search are more likely to be accurate because all documents in the same index represent a single entity.
- Each index can be sized appropriately for the number of documents it will contain: you can use a smaller number of primary shards for accounts and a larger number of primary shards for contacts.

## Upgrading from ElasticSearch 5.x.
- As both mapping types and the index names have changed, users will be required to run a full re-index before being able to use the new 7.x Engine.
- We will also need to either document for users or handle on upgrade the ```cache/ElasticSearchIndexer.lock``` file. If attempting to run a full re-index from a freshly upgraded instance that still contains the old cache, the system will instead launch a differential index (By design), normally this would be fine but as the user won't have any of the new indexes, elasticsearch will fail.

## How To Test This

1. Test ElasticSearch functionality using a 7.x ElasticSearch server.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [X] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [X] My change requires a change to the documentation.
- [X] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

File: lib/Search/ElasticSearch/ElasticSearchEngine.php [hunk 1]
```diff
@@ -88,7 +82,7 @@ public function search(SearchQuery $query): SearchResults
         $end = microtime(true);
         $searchTime = ($end - $start);
 
-        return new SearchResults($results, true, $searchTime, $hits['hits']['total']);
+        return new SearchResults($results, true, $searchTime, $hits['hits']['total']['value']);
```
Comment (clemente-raposo) [inline lib/Search/ElasticSearch/ElasticSearchEngine.php:85]: Hi @Dillon-Brown is this related with the index change? does seem so? Has the structure changed on elastic search?

File: lib/Search/ElasticSearch/ElasticSearchIndexer.php [hunk 2]
```diff
@@ -454,28 +464,11 @@ private function indexBatch($module, array $beans)
...
     {
         return $this->differentialIndexing && $this->lastRunTimestamp !== false;
     }
 
-    /**
-     * Retrieves the default params to set up an optimised default index for Elasticsearch.
-     *
-     * @return array
-     */
-    private function getDefaultMapParams()
```
Comment (clemente-raposo) [inline lib/Search/ElasticSearch/ElasticSearchIndexer.php:467]: Hi @Dillon-Brown, the reason for no longer needing the map params is related with elastic search 7?

**Review by clemente-raposo — APPROVED** [review]
Code looks ok. 

Added some comments to have a better understanding of the changes

File: lib/Search/ElasticSearch/ElasticSearchEngine.php (hunk 1 above)
Comment (Dillon-Brown) [reply lib/Search/ElasticSearch/ElasticSearchEngine.php:85]: Yes structure has changed on ElasticSearch, I've also changed what comes back from ```parseHits``` as it now needs to include ```module_source``` in it's array so that we can determine the index.

File: lib/Search/ElasticSearch/ElasticSearchIndexer.php (hunk 2 above)
Comment (Dillon-Brown) [reply lib/Search/ElasticSearch/ElasticSearchIndexer.php:467]: Yup exactly, ElasticSearch will now reject anything that comes with mappings like this.

**Review by Mac-Rae — DISMISSED** [review]

**Comment by Dillon-Brown** [comment]
Latest push is rebasing onto develop and a minor change to the ```composer.json```:

- Removed ```"elasticsearch/elasticsearch": "dist"``` and ```"rm -Rf vendor/elasticsearch/elasticsearch/tests/Elasticsearch/Tests"``` as the tests directory is no longer included.

**Review by clemente-raposo — APPROVED** [review]
latest changes look ok

**Comment by mattlorimer** [comment]
Build passing --> https://travis-ci.com/github/mattlorimer/SuiteCRM/builds/232272625

**Review by clemente-raposo — APPROVED** [review]
@Dillon-Brown latest changes look ok

**Comment by Mac-Rae** [comment]
Assessed :+1:

**Review by Mac-Rae — APPROVED** [review]
